[Mohamed Ammar Zumana Haseen] iP

amzhy opened the pull request:
## Introducing Duke!

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

Duke frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is, 

1. download it from [here](https://github.com/amzhy/ip/releases) 
2. double-click it
3. add your tasks
4. let it manage your tasks for you 👍 

And it is **FREE!**

### Features: 
- [X] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)

------------------------------------------------------------------------------------------------
If you are a Java programmer, you can use it to practise Java too! Here is the ```main``` method:

```Java
public class Duke {
    public static void main(String[] args) {
        new Duke().run();
    }
}
```

Commit:
Add Gradle support

amzhy committed:
Level-1: Greet, Echo, Exit

amzhy committed:
Level-2 Add, List

amzhy committed:
Level-3

amzhy committed:
Level-4

amzhy committed:
Level-5

amzhy committed:
Level-6

amzhy committed:
A-TextUiTesting

amzhy committed:
A-Enums

amzhy committed:
A-Enums

amzhy committed:
final edit

amzhy committed:
branch-level-8: date and time formats

amzhy committed:
branch-level-7: Add save feature

amzhy committed:
Merge remote-tracking branch 'origin/branch-level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/TaskList.java

amzhy committed:
Add date, time feature

amzhy committed:
Add more OOP

amzhy committed:
Refactor into packages

amzhy committed:
Add junit test

amzhy committed:
Add javadoc

amzhy committed:
Comply to coding standards

amzhy committed:
Level-9: find feature

amzhy committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/Ui.java

amzhy committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Storage.java
#	src/main/java/duke/Ui.java
#	src/main/java/duke/command/DeleteCommand.java
#	src/main/java/duke/command/DoneCommand.java

amzhy committed:
Fix bug for no txt file

junghon3709 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,113 @@
+ */
+public class Storage {
+
+    private static String filePath = "data/list.txt";
```
Maybe you should rename it to `FILE_PATH`? Since it's a constant

amzhy replied:
Yes :) I changed it to non-static so that I don't have to rename throughout the file. Thanks!

junghon3709 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,113 @@
+    }
+
+    /**
+     *
```
I think you're missing a method description?

junghon3709 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,113 @@
+     */
+    public ArrayList<Task> load() throws DukeException {
+        File f = new File(filePath);
+        ArrayList<Task> list = new ArrayList<>();
```
Maybe rename it to something more descriptive, such as `taskList`?

junghon3709 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,59 @@
+
+import duke.task.Task;
+
+import java.util.ArrayList;
```
I believe java imports should be placed above any other imports you have. I'm not sure if it's explictly stated in the Java coding standard, but it will show up in checkstyle if you do decide to implement it later on.

amzhy replied:
You're right, it caused a checkstyle error and has to precede the rest of the imports. Thanks!

junghon3709 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,46 @@
+    private static final String BYE_OUTPUT = "Bye. Hope to see you again soon!";
+    private static final String GREET_OUTPUT = "Hello! I'm Duke. What can I do for you?";
+    private static final String INDENT = "    ";
+    private static final String DIV_LINE = "    ____________________________________________________________________________";
```
Hmm, this line might be too long?

junghon3709 commented on src/main/java/duke/command/AddTaskCommand.java:
```diff
@@ -0,0 +1,85 @@
+/**
+ * This class adds different types of tasks
+ */
+public class AddTaskCommand implements  Command {
```
Extra spacing here?

junghon3709 commented on src/main/java/duke/command/AddTaskCommand.java:
```diff
@@ -0,0 +1,85 @@
+ * This class adds different types of tasks
+ */
+public class AddTaskCommand implements  Command {
+    private String type, detail;
```
I don't think declaring two variables in one line is allowed?

junghon3709 commented on src/main/java/duke/command/AddTaskCommand.java:
```diff
@@ -0,0 +1,85 @@
+        } else if (type.equals("deadline")) {
+            String[] description = detail.split("/by ", 2);
+            if (description.length == 1) {
+                throw new DukeException("Invalid Deadline entry.Try something like: deadline HW due /by 19/8/2021 14:00");
```
Lines might be too long in this file?

junghon3709 commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,24 @@
+import java.time.format.DateTimeFormatter;
+
+public class Event extends Task {
+    private LocalDateTime at;
```
Maybe change `at` to something more descriptive? I can think of `taskAt`?

junghon3709 commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,49 @@
+            int i = Integer.parseInt(text.trim()) - 1;
+            String t = taskList.removeTask(i);
+            ui.echo("Noted. I removed this task: " + t);
+        } catch (NumberFormatException | IndexOutOfBoundsException e) {
```
Wow, I didn't know you could catch multiple exceptions in the same line!

junghon3709 reviewed (commented):
Overall very clean code, just some small things to fix! :)

amzhy committed:
Merge remote-tracking branch 'origin/add-gradle-support'

xianlinc commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,46 @@
+        echo(e.toString());
+    }
+
+    public void echo(String next_input) {
```
Maybe next_input can be renamed to something else to follow camelCase?

xianlinc commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,29 @@
+import duke.DukeException;
+import duke.Storage;
+import duke.Ui;
+import duke.TaskList;
```
duke.Ui and duke.TaskList should be swapped to follow alphabetical order

xianlinc commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,36 @@
+
+public class Task {
+    private boolean isDone;
+    private String detail;
```
Maybe detail can be changed to something more descriptive

xianlinc reviewed (commented):
LGTM. Just some minor comments on naming of variables. Good code!

amzhy committed:
Use gradle to fix checkstyle

amzhy committed:
Use var args

rickyaandrew commented on data/list.txt:
```diff
@@ -0,0 +1,5 @@
+T | 1 | rit
+D | 1 | return bk | 19/12/2021 19:00
+D | 1 | do hw due | 19/08/2021 19:00
+E | 1 | funeral | 19/08/2021 23:00
+D | 0 | return t | 19/09/2021 19:00
```
data can be added into .gitignore

rickyaandrew commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,51 @@
+package duke;
+
+import duke.command.*;
+
+/**
+ * This class converts user inputs to Commands
+ */
+public class Parser {
+
+    /**
+     *
+     * @param cmd User input
+     * @return A Command to be executed
+     * @throws DukeException thrown if user input is invalid
+     */
```
JavaDocs is missing the descriptions

rickyaandrew commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,51 @@
+    /**
+     *
+     * @param cmd User input
+     * @return A Command to be executed
+     * @throws DukeException thrown if user input is invalid
+     */
+    public static Command parse(String cmd) throws DukeException {
+        String[] arr = cmd.split(" ", 2);
+        if (cmd.equals("bye")) {
+            return new ByeCommand();
+        } else if (cmd.equals("list")) {
+            return new ListCommand();
+        } else if (arr[0].equals("find")) {
+            if (arr.length == 1) {
+                throw new DukeException("Enter search keyword to find tasks");
+            } else {
+                return new FindCommand(arr[1]);
+            }
+        } else if (arr[0].equals("done")) {
+            if (arr.length == 1) {
+                throw new DukeException("Enter task no. to complete the task");
+            } else {
+                return new DoneCommand(arr[1]);
+            }
+        } else if (arr[0].equals("delete")) {
+            if (arr.length == 1) {
+                throw new DukeException("Enter task no. to delete the task");
+            } else {
+                return new DeleteCommand(arr[1]);
+            }
+        } else if (arr[0].equals("todo") || arr[0].equals("deadline") || arr[0].equals("event")) {
+            if (arr.length == 1 || arr[1].trim().length() == 0) {
+                throw new DukeException("Invalid Task entry. Description of a task cannot be empty.");
+            } else {
+                return new AddTaskCommand(arr);
+            }
+        } else {
+            throw new DukeException("Oops! Invalid entry.");
+        }
+    }
```
this is clean and neat, can try using switch() case method

rickyaandrew commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,14 @@
+package duke;
+
+public class DukeException extends Exception {
+    private final String error;
+
+    public DukeException(String error) {
+        this.error = error;
+    }
+
+    @Override
+    public String toString() {
+        return this.error;
+    }
+}
```
you can try going further by having other exceptions extending to duke exceptions, so that you wont be always be throwing only duke exceptions. maybe DukeStorageSaveException for exceptions faced while saving and DukeStorageLoadException for exceptions faced while loading data and files.

rickyaandrew commented on src/main/java/duke/command/ByeCommand.java:
```diff
@@ -0,0 +1,35 @@
+ */
+public class ByeCommand implements Command {
+
+    /**
+     *
+     * @param taskList Manages all current tasks
+     * @param ui Used to print messages
+     * @param storage Loads and saves the tasks to a txt file
+     * @throws DukeException thrown if there are input/parsing errors
+     */
```
noticed that most javadocs are not in a proper formatting. we have the same issue right here

rickyaandrew commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,49 @@
+     *
+     * @param s A string to be parsed to task number.
+     */
+    public DeleteCommand(String s) {
+        this.text = s;
+    }
```
usually its easier to read if

```
public DeleteCommand(String text) {
    this.text = text;
}
```

rickyaandrew reviewed (commented):
Overall the code is siccc and easy to read for readers. just that a little bit of clean ups need to be done. Good Job!

amzhy committed:
Add GUI

amzhy committed:
fix style errors

amzhy committed:
Improve GUI and fix JavaFX bugs

Long texts are truncated in labels.
Fixed bug by wrapping text.

GUI was improved to resemble a chatbot app
for better UX.

amzhy committed:
Add HelpCommand for all supported features

A help command reminds users about
relevant keywords to use for each feature.

amzhy committed:
Add assert checks

Assert checks added to TaskList and
DukeException.

amzhy committed:
Improve code quality

Rename method names for
better readability.

amzhy committed:
Use Streams feature

Use java streams to filter
tasks based on keyword.

amzhy committed:
Merge pull request #1 from amzhy/branch-A-Assertions

Add assert checks

amzhy committed:
Merge branch 'master' into branch-A-CodeQuality

amzhy committed:
Merge pull request #2 from amzhy/branch-A-CodeQuality

Improve code quality

amzhy committed:
Merge branch 'branch-A-Streams'

amzhy committed:
Add continuous integration

Use GitHub Actions to set up
Continuous Integration (CI).

amzhy committed:
Set theme jekyll-theme-midnight

amzhy committed:
Set theme jekyll-theme-minimal

amzhy committed:
Better GUI

amzhy committed:
Add screenshot

amzhy committed:
Update README.md

amzhy committed:
Set theme jekyll-theme-cayman

amzhy committed:
Update README.md

amzhy committed:
Fix links

amzhy committed:
Update userguide

amzhy committed:
Add javadoc

amzhy committed:
Update README.md

amzhy committed:
Fix screenshot
